File: Fast Exponentiation/FastExponentiation.java
import java.util.*;
import java.text.*;

class Power
{
    public static int calculatePower(int base, int exponent) {
        if (exponent == 0)
            return 1;

        if (exponent % 2 == 0)  // if exponent is even we just return sqaure of (a ^ n/2)
            return calculatePower(base, exponent / 2) * calculatePower(base, exponent / 2);

        else {  // if exponent is odd we just return a * sqaure of (a ^ n/2)
            return calculatePower(base, exponent / 2) * calculatePower(base, exponent / 2) * base;
        }
    }


    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter base: ");
        int a = scanner.nextInt();
        System.out.print("Enter exponent: ");
        int n = scanner.nextInt();

        System.out.println(MessageFormat.format("{0}^{1} = {2}", a, n, calculatePower(a, n)));
    }
}

File: Fast Exponentiation/fast_exponentiation.py
def calculate_power(base, power):
    if not power:
        return 1

    if not power % 2:
        return calculate_power(base, power // 2) * calculate_power(base, power // 2)

    else:
        return calculate_power(base, power // 2) * calculate_power(base, power // 2) * base


if __name__ == "__main__":
    a = int(input("Enter base: "))
    n = int(input("Enter exponent: "))
    print("{}^{} = {}".format(a, n, calculate_power(a, n)))